### ISSUE TITLE
Annotation extensions

### ISSUE THREAD

**Issue by pombase-admin** [pombase/canto#181]
Besic proposal for including annotation\_extensions:
https://sourceforge.net/apps/trac/pombase/wiki/IncludingAnnotationExtensions

Original comment by: ValWood

**Comment by kimrutherford** [comment]
The wiki page has [moved to GitHub](../wiki/AnnotationExtensions)

**Comment by kimrutherford** [comment]
I've done a bunch of code fixing in preparation for adding an annotation extension editing page.  Most of the changes aren't visible, but the workflow for adding an ontology annotation should be a bit snappier.

The main change is that there isn't a big pause after choosing the evidence, instead it immediately shows a summary of the annotation and the "comment" box.  Currently the annotation transfer isn't implemented but I will put that back if you prefer.  Is it still needed now that there is the option to duplicate annotations and change the gene/genotype?

**Comment by mah11** [comment]
I vote to scrap "annotation transfer" -- the "duplicate" option is better anyway.

**Comment by ValWood** [comment]
Yep, I agree, I thought that was the plan

**Comment by kimrutherford** [comment]
> Yep, I agree, I thought that was the plan

Just wanted to check :-)  Thanks.

**Comment by kimrutherford** [comment]
After help understanding ontology subsets (thanks Midori!), I'd like a new column to the configuration file "subset relation".  The idea is that we need to document which relations we follow to create the subsets.  I think the best place to configure that is in the config file rather than the code so I suggest these columns:

    term_ID
    term_name
    subset_relation
    allowed_extension_ids
    range
    text to plug in for tool

Midori suggests that we stick to is_a for creating subsets for now.

**Comment by kimrutherford** [comment]
Is there a configuration file ready that I can test?  I can use the example table from the wiki page for now: https://github.com/pombase/canto/wiki/AnnotationExtensions

**Comment by kimrutherford** [comment]
There is now a very basic annotation extension user interface that I can demo.  It's not in the test Canto yet as it's not wired into the rest of the ontology workflow.  I'm going to work on that now.  Hopefully it will be ready by Monday morning-ish.

The UI is very basic, but a lot of the behind the scenes stuff is working - there is now a script that will read the extension configuration table, pull out the subset ids (sub-ontologies) that are needed and store them in the Canto database.  The main Canto code can now parse and use the configuration file.  It uses that information and the subset information from the DB to populate the extension UI.

I've attached some screenshots of what's there so far.  Sorry for the random nonsense term IDs and descriptions.  The page is just for testing and so that we have a starting point for discussion.  It's probably not what you were thinking of, but now we have something to talk about it will be easier for me to understand what you'd like.

![screenshot 2015-09-06 17 47 16](https://cloud.githubusercontent.com/assets/90474/9702745/b6091a3a-54bf-11e5-8e23-2f9f6afdf95b.png)
![screenshot 2015-09-06 17 47 33](https://cloud.githubusercontent.com/assets/90474/9702746/b60ab318-54bf-11e5-8865-3fdb62c121aa.png)
![screenshot 2015-09-06 17 47 41](https://cloud.githubusercontent.com/assets/90474/9702747/b60b2b22-54bf-11e5-9110-80a9f851ff40.png)
![screenshot 2015-09-06 17 47 58](https://cloud.githubusercontent.com/assets/90474/9702749/b6113dd2-54bf-11e5-8ca0-d5ec5290091d.png)
![screenshot 2015-09-06 17 48 57](https://cloud.githubusercontent.com/assets/90474/9702748/b6103c16-54bf-11e5-992d-a050444f2eab.png)

**Comment by ValWood** [comment]
Lets chat on Monday morning.  Actually I can only do from about 10:15 ish...
We can see if Antonia can join us. Midori is at yeast conf in Italy.

**Comment by kimrutherford** [comment]
I can chat anytime until 1pm-ish on Monday.  Talk to you then.

**Comment by kimrutherford** [comment]
GO extensions from Val: https://www.dropbox.com/s/tckqmj1adljyc2k/go_annotation_extension_conf_val_email.txt?dl=0

**Comment by ValWood** [comment]
Antonia, Kim and I chatted.
We like what Kim has done so far :)

Next up:
1. Config files
Val will create config files for GO
Antonia will create config files for MOD
Midori will create config files for FYPO
Config files will live in
pombe-embl/supporting_files/annotation_ex_config
(format is in dropbox file but we don't need the term name column)

2. Interface
Display the annotation to the user as a human readable string in the same way that it looks on the gene page
i.e
has substrate Cdc2, during G2/Mtransition of mitotic cell cycle

3. By default we will assume that  the relations  used are mutually exclusive within a particular extension. Once  a relation has been used in a compound annotation, it will disappear from the available options. We will need to change this for the relations which can be used  multiply within an extension, and we will flag these in some way in the config file (open a ticket  with the details and examples as identified).

4. When an AE annotation is completed, there will be 2 options." Proceed", or "add another extension"

Was that it?

**Comment by ValWood** [comment]
Allow this:

Some enzymes act on a specific(defined) subset of proteins.

For example
GO:0032542 sulfiredoxin activity
peroxiredoxin-(S-hydroxy-S-oxocysteine) + ATP + 2 R-SH = peroxiredoxin-(S-hydroxycysteine) + ADP + phosphate + R-S-S-R

where peroredoxin =  PomBase:SPBC32F12.03c|PomBase:SPBC1773.02c|PomBase:SPCC330.06c|PomBase:SPCC576.03c|PomBase:SPAC27D7.12c|PomBase:SPBC17D11.03c|PomBase:SPBC17D11.03c

so, in this case, I have put 
GO:0032542 PomBase:SPBC32F12.03c|PomBase:SPBC1773.02c|PomBase:SPCC330.06c|PomBase:SPCC576.03c|PomBase:SPAC27D7.12c|PomBase:SPBC17D11.03c|PomBase:SPBC17D11.03c sulfiredoxin for

(you can default this to PomBase GENE until you are ready to deal with ranges)

**Comment by ValWood** [comment]
I have used the following (easy to change)

PomBase:ProteinID
PomBase:tRNAID
PomBase:TranscriptID 
to specify a protein, tRNA and transcript respectively.

Kim, for now you can default all to gene, I just wanted to capture the different things and we will almost certainly want to change the labels. At the moment we don't even use transcript IDs, we just 'assume' if its a term which acts on a transcript that the thing in the extension is referring to the transcript.

Anyways, I just wanted to let you know what I had done, and I'll fix to the consensus when there is one.

**Comment by ValWood** [comment]
In the GO config  files detailed here:
https://github.com/pombase/curation/issues/684#issuecomment-138809157

When you start to use the 'range' info for the  IDs allowed in the extensions, where I put a GO or SO ID in the range column it means this ID and all of its descendants (I guess that is obvious). There might be some duplications of the SO terms where ancestors and descendants are both included. We can tweak this later (but it probably doesn't really matter, as it is only  the 'list of terms which will be made  available via the autocomplete?).

I can't think of anything else you need to know. It should be just a case of curators refining the files iteratively until the biology is good, but I think this should include ~99%% of  current GO extension use cases.

**Comment by mah11** [comment]
Re this from above:

> By default we will assume that the relations used are mutually exclusive within a particular extension. Once a relation has been used in a compound annotation, it will disappear from the available options. We will need to change this for the relations which can be used multiply within an extension, and we will flag these in some way in the config file (open a ticket with the details and examples as identified).

It already doesn't make sense to implement "once a relation has been used ... it will disappear from the available options" globally, because every FYPO "protein binding" phenotype annotation requires two assayed_using(PomBase:proteinID) extensions (for binding between x and y). At present it's enforced at the level of a warning in the chado_checks log.

**Comment by ValWood** [comment]
That's what this menas

> We will need to change this for the relations which can be used multiply within an extension, and we will flag these in some way in the config file (open a ticket with the details and examples as identified).

There will be exceptions for protein binding terms....

We can do it another way, but we need to be able to prevent people from selecting the same relation/extension twice to make a compound annotation with a term where it should not be possible.

**Comment by mah11** [comment]
Are we going to need a cardinality column in the config files?

**Comment by ValWood** [comment]
For? (Sorry being dim).

**Comment by mah11** [comment]
A cardinality column would specify how many of a given relation would be allowed per extension. Most things would be 0 (because not mandatory) or 1 (only use once per extension); FYPO protein binding would be 2 (two required; not allowed to skip it).

**Comment by ValWood** [comment]
Ah right. I think we probably do....

**Comment by mah11** [comment]
I've created two annotation extension configuration files for FYPO. The "admin-only" one has a few domain/range combinations that should only be available to admins/professional curators/us. Between them the two files should cover all the extensions I found in the v54 PHAF, except the ones in curation ticket 689 https://github.com/pombase/curation/issues/689 (open another ticket if you spot anything I missed).

pombe-embl/supporting_files/annotation_ex_config/FYPO_A_E_admin-only_config
pombe-embl/supporting_files/annotation_ex_config/FYPO_A_E_config

**Comment by mah11** [comment]
oh yeah, I've also taken the liberty of adding a column for cardinality, and a header row; can delete if they break anything

**Comment by kimrutherford** [comment]
> oh yeah, I've also taken the liberty of adding a column for cardinality

Thanks.

> and a header row; can delete if they break anything

A header row is fine as long as it's in all files and identical in all files.

**Comment by mah11** [comment]
> A header row is fine as long as it's in all files and identical in all files.

Well, it should be, 'cos I copied it from one file and pasted it into the others!

**Comment by kimrutherford** [comment]
There is now a basic annotation extension GUI for testing.  The configuration files from pombe-embl are loaded in the test Canto and it's using the new code.  The current code is just for testing.  It will need changing after we chat about it.

The annotation edit dialog now shows the extensions with an edit button that sort of works.  Only extensions adding using the new interface in the test tool will be visible and editable at the moment.  The older free text extensions need to be upgraded to fit in the new system.  After I've fixed a few more obvious problems, I'll close this issue and we can open an separate issue for each change needed to get this ready for the community.

![screenshot 2015-10-08 17 26 14](https://cloud.githubusercontent.com/assets/90474/10357726/f8c20dce-6de1-11e5-85c1-5d4c3f14add5.png)

![screenshot 2015-10-08 17 36 42](https://cloud.githubusercontent.com/assets/90474/10357822/2e1c8494-6de3-11e5-896f-84ce42a7b144.png)

**Comment by ValWood** [comment]
Lets all have a little test today if poss (I think I'll be off Fri, I ended up working yesterday), and give general feedback.
We'll have group&curator call Tues morning to  catch up on stuff from over the past 2 weeks.
Will arrange a call with Alistair for soon after that.

**Comment by ValWood** [comment]
I think theres a bug, the options don't show

**Comment by ValWood** [comment]
![no_options](https://cloud.githubusercontent.com/assets/7359272/10359808/dbf15f90-6d93-11e5-82c2-acf9528890c0.jpg)

**Comment by kimrutherford** [comment]
Sorry about that.  It worked when I tried it early.  Not a great start.

**Comment by kimrutherford** [comment]
Should be fixed now.  Could you try again?  ATPase activator activity is a good test

**Comment by ValWood** [comment]
It works but I was only propmted for 'happens_during'
Sould this have two avaiable extensions (enzyme activators you can specifiy the actual ATPAse substrate, I think?)

Also, you might not have done this part yet, but  happens_during should allow a BP (and evnetually a subset of BP) in the extension (currently it is MF)

**Comment by ValWood** [comment]
other than that its working.
You can now get rid of the follow up box to add free text extensions? in cases where the prompts are present (Q should we keep it for the temi being for  others to pick up stragglers?)

Also did we decide how we were going to do the looping for additional extensions? and make it clear to the user when it was a compound extension? I can't remember? I know we discussed it...

**Comment by kimrutherford** [comment]
I got two available extensions in this session with ATPase activator activity:
  http://curation.pombase.org/test/curs/35bfa0019d25c550/feature/gene/annotate/1/start/molecular_function

What session were you using?

> Also, you might not have done this part yet, but happens_during should allow a BP
> (and evnetually a subset of BP) in the extension (currently it is MF)

The configuration file just has MF:
GO:0003674      is_a    happens_during  GO:0003674      has function during     0,1

> You can now get rid of the follow up box to add free text extensions?

There shouldn't be any free text extension box now.  Where are you seeing that?

> I know we discussed it...

We had some ideas I think but didn't make a decision.

**Comment by ValWood** [comment]
O'm using
http://curation.pombase.org/test/curs/8a16b8d3a732e18b/feature/gene/view/1
and its fine. MAybe I was using the wrong term

re 
GO:0003674 is_a happens_during GO:0003674 has function during 0,1
I need to fix GO:0003674 to GO:0008150 biological_process
Do I edit the files in  svn or did you move them to github already?

>There shouldn't be any free text extension box now. Where are you seeing that?
After I make the annotation and hit proceed

**Comment by ValWood** [comment]
As a start, we could edit this text
These are the possible extension types for GO:0001671, please add as many as are appropriate:

to something like:

"Please add as many as are appropriate to make a compound annotation (all occur simultaneously).
You will be given the option to  make additional independent extensions"

And then just add the "add another extension" button (didn't we already  decide on some  shorter text for this button"?) next to "proceed" button?

**Comment by kimrutherford** [comment]
> Do I edit the files in svn

They are in SVN: pombe-embl/supporting_files/annotation_ex_config/

> After I make the annotation and hit proceed

There's a general comment box.  Is that what you mean?  I thought that was still useful.  Did I misunderstand?

**Comment by ValWood** [comment]
Yes the general comment box is useful. Sorry I only skim read it and I thought it mentioned extensions.
Val

**Comment by kimrutherford** [comment]
> As a start, we could edit this text

I've changed that.

**Comment by ValWood** [comment]
We will probably tweak it...

**Comment by mah11** [comment]
Now I've had a brief go, and I'm seeing the bug Val reported last week.

I fired up a new session for testing (http://curation.pombase.org/test/curs/892645d1cd461bbf/), but when I try to add extensions to any annotation (have tried GO MF, GO BP and FYPO so far) I get the text but no box, pulldown, etc. to add an extension.

I also see the same thing in Kim's session (35bfa0019d25c550).

**Comment by mah11** [comment]
I put some suggestions in #988 and #989 for text and workflow/looping/finishing.

**Comment by ValWood** [comment]
Looks good. Any ideas why the above didn't spawn notifications?

**Comment by ValWood** [comment]
We can discuss tomorrow. I'm only "sort of" working today. feeling crap so went back to bed. just doing a few e-mails in between dozing.
Hopefully will be better tomorrow.

**Comment by mah11** [comment]
Transferred bug to #992.

Closing as agreed on call this morning. Shiny new tickets for shiny new bugs from now on.

**Comment by kimrutherford** [comment]
> Now I've had a brief go, and I'm seeing the bug Val reported last week.

Sorry about that.  I was confused because it had worked earlier for me but I've realised the problem.

The terms in the test database have some extra data attached to allow the extension GUI to quickly look up parents that are domains of extension relations (so we don't need to load the full closure).  I forgot that the nightly ontology update works by clearing the existing terms and replacing them with the new OBO file.  It also clears the data for the extension parent terms so the extension GUI thought there was nothing to show.

So I loaded the extension stuff yesterday and it worked then it got wiped mid afternoon my time.  Whoops.

Currently the extra properties for the extensions are added by a script I run manually as needed.  I'll move that function into the ontology load code so it happens seamlessly each night.

(And I'll run the script now to make sure it works for you tomorrow)
